docs/framework-overview.md

```diff
@@ -0,0 +1,398 @@
+
+The installation process in the extension step starts in
+`EasyBlock.install_extensions`, which will determine the installation approach
+for the extensions (sequential or parallel based on given options/support) and
+call the `Extension.run` method of each extension instance. This `run` method
+is the one carrying out the actual installation of the extension and typically
+executes the install step of the parent `EasyBlock` with some extra actions
+before and/or after it. Nonetheless, custom easyblocks for extensions can fully
+redefine their `run` method.
+
+!!! warning
+    In EasyBuild v5.0 the name of the `Extension.run` method will be changed
+    to a more meaningful and less generic name.
+
+Framework also provides the `ExtensionEasyBlock` class which inherits from both
+`EasyBlock` and `Extension`. This class is useful for software that can be
+either installed standalone with their own easyconfig file or as an extension:
+
+* as extension: `ExtensionEasyBlock` uses the `Extension` class
+
+* stand-alone: `ExtensionEasyBlock` uses the `EasyBlock` class
+
+For instance, `PythonPackage`, `RPackage` or `PerlModule` are generic easyblocks
+commonly used to install standalone packages or extensions of packages that use
+other easyblocks. You will find many examples of both cases in our
+[easyblock-easyconfigs](https://github.com/easybuilders/easybuild-easyconfigs/)
+repository. 
+
+## The EasyConfig class
+
+The EasyConfig class handles parsing and validating of easyconfig files. As
+such, EasyConfig is independent from EasyBlock and it is actually needed to
+determine the EasyBlock used in the installation of the target software
+package.
+
+Once the target easyconfig file is read and all its parameters are validated,
+they will be accessible through different attributes and methods of the
+EasyConfig object. The main examples are:
+
+* `EasyConfig.name` returns the name of the package
+
+* `EasyConfig.version` returns the version of the package
+
+* `EasyConfig.toolchain` returns the toolchain used by the package
+
+* `EasyConfig.dependencies()` returns the list of dependencies
+
+!!! note
+    The EasyConfig instance is also accessible from the corresponding EasyBlock
+    instance through its `EasyBlock.cfg` attribute.
+
+## Toolchain mechanism
+
+A [toolchain][toolchains] in EasyBuild is much more than just a bundle of
+libraries and compilers to build and run software in a consistent environment.
+Toolchains are defined by their own classes built on top of the `Toolchain`
+class.
+
+The base `Toolchain` class is defined in `easybuild.tools.toolchain.toolchain`.
+It does not only provide the attributes defining the toolchain (*i.e.* name and
+version) but also several methods to interact with it:
+
+* set the environment of the toolchain: `Toolchain.prepare`
+
+    * validate dependencies for the toolchain: `Toolchain._check_dependencies`
+
+    * handle build options: `Toolchain.prepare_rpath_wrappers`,
```

Comment:
These two should not be indented?

Reply:
As discussed, these points happen inside `prepare`, so that's why they are indented.